### ISSUE TITLE
`torch.export` ViT+flash attention:  `Attempting to use FunctionalTensor on its own. Instead, please use it with a corresponding FunctionalTensorMode`

### ISSUE THREAD

**Issue by bhack** [pytorch/pytorch#140400]
### 🐛 Describe the bug

Exporting a VIT with flexattention I got this error https://github.com/pytorch/pytorch/issues/137759#issuecomment-2470595683

As in that thread It seems that is a distinct issue I've opened this one.

### Versions

nightly

**Comment by bdhirsh** [comment]
@bhack any chance you have some repro code to run? (or at the least a full stacktrace? The one in the linked comment looks incomplete)

**Comment by bhack** [comment]
>  any chance you have some repro code to run? (or at the least a full stacktrace? The one in the linked comment looks incomplete)

How we could have a repro code if this just emerge on `torch.export` over a trained model/checkpoint with flex attention.

I've not omitted anything interesting from the export log:
```python
 Attempting to use FunctionalTensor on its own. Instead, please use it with a corresponding FunctionalTensorMode()
While executing %dense_mask : [num_users=2] = call_method[target=new_zeros](args = (%_add_batch_dim_3, 1, 2), kwargs = {dtype: torch.int32})
Original traceback:
  File "/workspace/modeling/meta_arch/vitmatte.py", line 34, in forward
    features = self.backbone(inputs)
  File "/workspace/modeling/backbone/vit.py", line 504, in forward
    x = blk(x)
  File "/workspace/modeling/backbone/vit.py", line 362, in forward
    x = self.attn(x)
  File "/workspace/modeling/backbone/vit.py", line 77, in forward
    attn_out = flex_attention(
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 1197, in flex_attention
    block_mask = _create_empty_block_mask(query, key)
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 893, in _create_empty_block_mask
    return BlockMask.from_kv_blocks(
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 349, in from_kv_blocks
    q_num_blocks, q_indices = _transpose_ordered(kv_num_blocks, kv_indices)
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 187, in _transpose_ordered
    dense = _ordered_to_dense(num_blocks_in_row, col_indices)
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 172, in _ordered_to_dense
    out = create_dense_batched(num_blocks_in_row, col_indices)
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/apis.py", line 203, in wrapped
    return vmap_impl(
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/vmap.py", line 331, in vmap_impl
    return _flat_vmap(
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/vmap.py", line 479, in _flat_vmap
    batched_outputs = func(*batched_inputs, **kwargs)
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/apis.py", line 203, in wrapped
    return vmap_impl(
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/vmap.py", line 331, in vmap_impl
    return _flat_vmap(
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/vmap.py", line 479, in _flat_vmap
    batched_outputs = func(*batched_inputs, **kwargs)
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 153, in create_dense_one
    dense_mask = kv_indices.new_zeros(num_rows, num_cols + 1, dtype=torch.int32)
 ```

**Comment by bdhirsh** [comment]
@bhack that stacktrace shows the "Original Stacktrace", but I was also expecting to see the stacktrace inside of the export code itself. E.g. for the code that raises this error:
```
Attempting to use FunctionalTensor on its own. Instead, please use it with a corresponding FunctionalTensorMode()
```

I don't see the actual stacktrace that invoked that line (https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/functional_tensor.py#L221)

**Comment by bhack** [comment]
I've suppressed the try/catch we have few more lines:
```python
success = export_model(
              ^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/export_model_export.py", line 125, in export_model
    exported_program = export(model, (example_inputs,), dynamic_shapes=dynamic_shapes)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/export/__init__.py", line 368, in export
    return _export(
           ^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/export/_trace.py", line 1031, in wrapper
    raise e
  File "/opt/conda/lib/python3.11/site-packages/torch/export/_trace.py", line 1004, in wrapper
    ep = fn(*args, **kwargs)
         ^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/export/exported_program.py", line 122, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/export/_trace.py", line 1957, in _export
    export_artifact = export_func(  # type: ignore[operator]
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/export/_trace.py", line 1251, in _strict_export
    return _strict_export_lower_to_aten_ir(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/export/_trace.py", line 1360, in _strict_export_lower_to_aten_ir
    aten_export_artifact = lower_to_aten_callback(
                           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/export/_trace.py", line 738, in _export_to_aten_ir
    gm, graph_signature = transform(aot_export_module)(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/aot_autograd.py", line 1278, in aot_export_module
    fx_g, metadata, in_spec, out_spec = _aot_export_function(
                                        ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/aot_autograd.py", line 1517, in _aot_export_function
    fx_g, meta = create_aot_dispatcher_function(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/aot_autograd.py", line 527, in create_aot_dispatcher_function
    return _create_aot_dispatcher_function(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/aot_autograd.py", line 778, in _create_aot_dispatcher_function
    compiled_fn, fw_metadata = compiler_fn(
                               ^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py", line 113, in aot_dispatch_export
    graph, _, _ = aot_dispatch_base_graph(
                  ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py", line 152, in aot_dispatch_base_graph
    fw_module = _create_graph(
                ^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py", line 55, in _create_graph
    fx_g = make_fx(
           ^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py", line 2172, in wrapped
    return make_fx_tracer.trace(f, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py", line 2110, in trace
    return self._trace_inner(f, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py", line 2081, in _trace_inner
    t = dispatch_trace(
        ^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_compile.py", line 32, in inner
    return disable_fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_dynamo/eval_frame.py", line 721, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py", line 1137, in dispatch_trace
    graph = tracer.trace(root, concrete_args)  # type: ignore[arg-type]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py", line 1677, in trace
    res = super().trace(root, concrete_args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_dynamo/eval_frame.py", line 721, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/_symbolic_trace.py", line 843, in trace
    (self.create_arg(fn(*args)),),
                     ^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py", line 1192, in wrapped
    out = f(*tensors)
          ^^^^^^^^^^^
  File "<string>", line 1, in <lambda>
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/_aot_autograd/traced_function_transforms.py", line 693, in inner_fn
    outs = fn(*args)
           ^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/_aot_autograd/traced_function_transforms.py", line 413, in _functionalized_f_helper
    f_outs = fn(*f_args)
             ^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/_aot_autograd/traced_function_transforms.py", line 78, in inner_fn
    outs = fn(*args)
           ^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/_aot_autograd/utils.py", line 184, in flat_fn
    tree_out = fn(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/_aot_autograd/traced_function_transforms.py", line 875, in functional_call
    out = PropagateUnbackedSymInts(mod).run(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/interpreter.py", line 167, in run
    self.env[node] = self.run_node(node)
                     ^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/experimental/symbolic_shapes.py", line 6570, in run_node
    result = super().run_node(n)
             ^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/interpreter.py", line 228, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/interpreter.py", line 332, in call_method
    return getattr(self_obj, target)(*args_tail, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py", line 1240, in __torch_function__
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py", line 1287, in __torch_function__
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_ops.py", line 840, in handler
    return torch._library.utils.handle_dispatch_mode(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_library/utils.py", line 295, in handle_dispatch_mode
    return curr_mode.__torch_dispatch__(op_overload, overload_types, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_subclasses/functional_tensor.py", line 545, in __torch_dispatch__
    outs_unwrapped = func._op_dk(
                     ^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_ops.py", line 840, in handler
    return torch._library.utils.handle_dispatch_mode(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_library/utils.py", line 295, in handle_dispatch_mode
    return curr_mode.__torch_dispatch__(op_overload, overload_types, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/utils/_stats.py", line 21, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py", line 1342, in __torch_dispatch__
    return proxy_call(self, func, self.pre_dispatch, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py", line 911, in proxy_call
    out = func(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_ops.py", line 723, in __call__
    return self._op(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/_subclasses/functional_tensor.py", line 221, in __torch_dispatch__
    raise RuntimeError(
RuntimeError: Attempting to use FunctionalTensor on its own. Instead, please use it with a corresponding FunctionalTensorMode()

While executing %dense_mask : [num_users=2] = call_method[target=new_zeros](args = (%_add_batch_dim_3, 1, 2), kwargs = {dtype: torch.int32})
Original traceback:
  File "/workspace/modeling/meta_arch/vitmatte.py", line 34, in forward
    features = self.backbone(inputs)
  File "/workspace/modeling/backbone/vit.py", line 504, in forward
    x = blk(x)
  File "/workspace/modeling/backbone/vit.py", line 362, in forward
    x = self.attn(x)
  File "/workspace/modeling/backbone/vit.py", line 77, in forward
    attn_out = flex_attention(
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 1197, in flex_attention
    block_mask = _create_empty_block_mask(query, key)
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 893, in _create_empty_block_mask
    return BlockMask.from_kv_blocks(
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 349, in from_kv_blocks
    q_num_blocks, q_indices = _transpose_ordered(kv_num_blocks, kv_indices)
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 187, in _transpose_ordered
    dense = _ordered_to_dense(num_blocks_in_row, col_indices)
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 172, in _ordered_to_dense
    out = create_dense_batched(num_blocks_in_row, col_indices)
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/apis.py", line 203, in wrapped
    return vmap_impl(
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/vmap.py", line 331, in vmap_impl
    return _flat_vmap(
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/vmap.py", line 479, in _flat_vmap
    batched_outputs = func(*batched_inputs, **kwargs)
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/apis.py", line 203, in wrapped
    return vmap_impl(
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/vmap.py", line 331, in vmap_impl
    return _flat_vmap(
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/vmap.py", line 479, in _flat_vmap
    batched_outputs = func(*batched_inputs, **kwargs)
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 153, in create_dense_one
    dense_mask = kv_indices.new_zeros(num_rows, num_cols + 1, dtype=torch.int32)
```

**Comment by drisspg** [comment]
@bhack  can you try creating a standalone repro that we can run?

It is very hard for us to look at stack traces and come up with fixes without building some sort of test loop. I understand that some of this code may be proprietary but anything you can do on your end to make debugging easier will allow us to fix the problem faster

**Comment by bhack** [comment]
Any suggestions as it just happened at the export of a private checkpoint that I cannot share. 
The training is ok.

**Comment by drisspg** [comment]
cc @bdhirsh  for ideas but maybe a small representative nn.module that calls flex_attention the same way as in your E2E code

**Comment by bhack** [comment]
I have opened a lot of tickets in the last few years and I am starting to think more and more  that we need to find an equilibrium. I know it is the same request at every ticket to have a minimal source code repro but to re-create a minimal source repo when you work e2e for every ticket is a huge amount of time also for the reporter.

The problem is that an user or potential  contributor needs to understand what an error like  `Attempting to use FunctionalTensor on its own. Instead, please use it with corresponding FunctionalTensorMode()`  really means and eventually localize a PR  without studying all the almost undocumented internals.

This requires feedback and pointers from core developers. Also when we have this often the review on  external PRs are starting to be extremely slow (see https://github.com/pytorch/pytorch/pull/136364) and this disencitivate external contribution a bit.

I think users with potential contributions want to spend time co-investigate and eventually contribute to fix errors more than spend time in trial and error to reproduce a minimal code over an end2end model as the user often have the advantage to have the full source code also when this could be not shared.

**Comment by drisspg** [comment]
@bhack I agree with the sentiment. Are you saying that you would rather fix and contribute then create minimal reproducing code for PyTorch Devs. I think the unfortunate reality is that even for PyTorch core developers it is often really hard to disambiguate the many subsytems of PyTorch in order to isolate problems.

I think the main reason why often urge issue creators (and contributors) to create small reproducible code samples is because if we had access to the full source code and were debugging this is the very thing we would do. Especially since we would ideally like to add Unit Tests that can ensure regressions dont happen.   I let @bdhirsh  speak to the specific in this example

**Comment by bhack** [comment]
I fully understand this,  also in TF few years ago, I was a promoter to contribute eventually only failing test from issues to then be contributed or fixed later (so expected failing test). 
This cause often we have orphan users issues accumulation over the time (for months or years) that are instead silently solved by nighties or new releases.

The points in my experience with the project are:
- How much time the user spent reproducing the issue from E2E to standalone code. Often with the compiler stack issues it is going to require a lot of time cause we are tracing the chain and so we need to reproduce eventually all the leaf inputs to the failing isolated function.
- Users could do some steps on its own side if guided by the core team cause they have the full source code to investigate without allocating a lot of time to the standalone minimal reproduction. Sometimes this could be faster with the right feedback then with extensive trial and error to extract a minimal standalone repro.
- If you really want to invest time for a PR, after few guided feedback, you are really going to evaluate how the review rounds are becoming too diluted. Cause if you have to change the working context too many times over months you will really start to evaluate if there is a value  to contribute something or not to the project rather than continue to follow your daily activities and do a refactoring that ultimately ignores the problem itself.

Also I understand that skilled internal devels that are working every days on the internals and with off the channel discussion support are faster to fix internals then external users. But at some point we always finding that there is a bandwidth limit if we don't want to invest time to let the community growth.
So it's a balance between different constraints and goals but let me know if there's a better place where we can go deeper into the this thread as it is a quite general and I don't want to go off-topic here.

**Comment by bhack** [comment]
So in the meantime:
- It seems we don't have any `torch.export` test with `flex_attention` right? If true we don't have test coverage at all for exporting `flex_attention`

On this specific case, here a repro:
```python
import torch
from torch import nn, Tensor
from torch.export import export, Dim
from torch.nn.attention.flex_attention import flex_attention

class MinimalAttention(nn.Module):
    """Minimal reproduction of Attention layer using flex_attention."""
    
    def __init__(self, dim, num_heads=8, qkv_bias=True):
        super().__init__()
        self.num_heads = num_heads
        self.head_dim = dim // num_heads
        self.scale = self.head_dim ** -0.5

        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
        self.proj = nn.Linear(dim, dim)

    def forward(self, x: Tensor) -> Tensor:
        B, N, C = x.shape  # Flattened spatial dimensions with batch
        qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, self.head_dim).permute(0, 3, 2, 1, 4)
        q, k, v = qkv[:, :, 0], qkv[:, :, 1], qkv[:, :, 2]

        # Compute attention output using flex_attention
        attn_out = flex_attention(q, k, v, scale=self.scale)
        return self.proj(attn_out.transpose(1, 2).reshape(B, N, C))

def replicate_error():
    B, N, C = 2, 1024, 64
    x = torch.ones(B, N, C, device="cuda", dtype=torch.float32)

    # Initialize the minimal attention model
    model = MinimalAttention(dim=C).cuda()

    # Set up dynamic shape for batch size flexibility
    dynamic_shapes = {
        "x": {0: Dim("batch", min=1, max=B), 1: Dim.STATIC, 2: Dim.STATIC}
    }

    # Export with AMP and dynamic shape settings
    try:
        with torch.amp.autocast(device_type="cuda", dtype=torch.float16):
            exported_model = export(model, (x,), dynamic_shapes=dynamic_shapes)
        torch.save(exported_model, "minimal_attention_error_model.pt")
        print("Exported model successfully.")
    except RuntimeError as e:
        print("Export failed with error:", e)

# Run the minimal reproduction
replicate_error()
```

```python
Export failed with error: Attempting to use FunctionalTensor on its own. Instead, please use it with a corresponding FunctionalTensorMode()
While executing %dense_mask : [num_users=2] = call_method[target=new_zeros](args = (%_add_batch_dim_3, 1, 2), kwargs = {dtype: torch.int32})
......
File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 153, in create_dense_one
    dense_mask = kv_indices.new_zeros(num_rows, num_cols + 1, dtype=torch.int32)
```

**Comment by tugsbayasgalan** [comment]
Export is actually moving away from functional pre-dispatch IR and the move is expected to happen very soon (in a week). As a result torch.export.export_for_training should be an alias to export moving forward, so you can temporarily use export_for_training to unblock.

**Comment by bhack** [comment]
> Export is actually moving away from functional pre-dispatch IR and the move is expected to happen very soon (in a week). As a result torch.export.export_for_training should be an alias to export moving forward, so you can temporarily use export_for_training to unblock.

Do you mean also if my export is for inference only?

**Comment by bhack** [comment]
**OFF-TOPIC note** @drisspg as you see in this case the reproduction was really not required.

**Comment by bhack** [comment]
@tugsbayasgalan After using `export_for_inference` there is another export issue around the corner that instead is not visible during training.
```python
failed with AMP: NYI: querying is_contiguous inside of vmap for memory_format other than torch.contiguous_format

While executing %dense_mask : [num_users=2] = call_method[target=new_zeros](args = (%_add_batch_dim_3, 1, 2), kwargs = {dtype: torch.int32})
Original traceback:
  File "/workspace/modeling/meta_arch/vitmatte.py", line 34, in forward
    features = self.backbone(inputs)
  File "/workspace/modeling/backbone/vit.py", line 504, in forward
    x = blk(x)
  File "/workspace/modeling/backbone/vit.py", line 362, in forward
    x = self.attn(x)
  File "/workspace/modeling/backbone/vit.py", line 77, in forward
    attn_out = flex_attention(
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 1197, in flex_attention
    block_mask = _create_empty_block_mask(query, key)
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 893, in _create_empty_block_mask
    return BlockMask.from_kv_blocks(
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 349, in from_kv_blocks
    q_num_blocks, q_indices = _transpose_ordered(kv_num_blocks, kv_indices)
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 187, in _transpose_ordered
    dense = _ordered_to_dense(num_blocks_in_row, col_indices)
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 172, in _ordered_to_dense
    out = create_dense_batched(num_blocks_in_row, col_indices)
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/apis.py", line 203, in wrapped
    return vmap_impl(
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/vmap.py", line 331, in vmap_impl
    return _flat_vmap(
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/vmap.py", line 479, in _flat_vmap
    batched_outputs = func(*batched_inputs, **kwargs)
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/apis.py", line 203, in wrapped
    return vmap_impl(
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/vmap.py", line 331, in vmap_impl
    return _flat_vmap(
  File "/opt/conda/lib/python3.11/site-packages/torch/_functorch/vmap.py", line 479, in _flat_vmap
    batched_outputs = func(*batched_inputs, **kwargs)
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 153, in create_dense_one
    dense_mask = kv_indices.new_zeros(num_rows, num_cols + 1, dtype=torch.int32)
```

**Comment by bhack** [comment]
@zou3519 Do you think it is connect to internal implementation of flex_attention vmap as in https://github.com/pytorch/pytorch/issues/91437?

But why it is going to emerge only on export?

**Comment by bhack** [comment]
New ticket for `export_for_inference` at https://github.com/pytorch/pytorch/issues/140849

**Comment by zou3519** [comment]
@bhack do you have more context over what you're doing? Let's say we are able to get torch.export on ViT+flex attention to work. What are you planning on doing with the exported program?

**Comment by bhack** [comment]
@zou3519 You have the rational for this error in @tugsbayasgalan's reply at https://github.com/pytorch/pytorch/issues/140400#issuecomment-2479457579.

Then for `export_for_inference` see the full reproduction at https://github.com/pytorch/pytorch/issues/140849

**Comment by bdhirsh** [comment]
> as you see in this case the reproduction was really not required.

This can be tough: I agree with @drisspg that repros are extremely helpful (and pretty frequently are necessary in order to root cause), because they mimic what developers would probably do to diagnose the issue anyway - most of the bugs that I've looked into, at least, have been the result of an edge-case interaction between several subsystems, that is usually hard to guess at from the error message alone. I agree that in this particular case though, it looks like the interaction was simple enough that I / someone could have constructed the failure just from trying those features together from first principles.

**Comment by bhack** [comment]
@bdhirsh Yes that comment was specific to this case or similar but in general I have exposed many other valid points in the previous comment at https://github.com/pytorch/pytorch/issues/140400#issuecomment-2474903378

I hope we could find a better place to discuss these points as here we are at risk to go off-topic.

**Comment by zou3519** [comment]
> @zou3519 You have the rational for this error in @tugsbayasgalan's reply at [#140400 (comment)](https://github.com/pytorch/pytorch/issues/140400#issuecomment-2479457579).
> 
> Then for `export_for_inference` see the full reproduction at #140849

@bhack I'm interested in what you're planning on doing with the exported program, assuming we do resolve this issue. Because there may be other problems depending on your use case

**Comment by bhack** [comment]
I want to export for inference so you can comment directly at https://github.com/pytorch/pytorch/issues/140849

As it is clear what is instead the problem here.

**Comment by drisspg** [comment]
Looking through old issues: @tugsbayasgalan, Any update here?

**Comment by jansel** [comment]
My agent was unable to reproduce this on 5d72a01abe3c335340bffb58cf23e5e244c521d6. Reopen if you see this on the latest PyTorch nightly
